### PR TITLE
Improve Caliper/Profiling support

### PR DESCRIPTION
**Pull request by ndryden**
This improves our support for Caliper, and for profiling in general:
* The `prof_stop` function is now actually called by the profiler callback, so profilers can clean up.
* Adiak takes a dup of our world comm so it can gather MPI information.
* Caliper is now configurable via a new `--caliper_config` command-line option.
* The Python frontend now has support for `--profile`, `--profile-init`, and `--caliper-config` options, which do what you think they do.
* The ResNet, CosmoFlow, and Transformer applications have been updated to make use of these.

### CODE REVIEW

**Review by benson31 — APPROVED** [review]
:shipit:

File: applications/nlp/transformer/train.py [hunk 1]
```diff
@@ -226,8 +226,9 @@ def make_batch_script(trainer_params, model_params, script_params, args):
             lbann.CallbackProgressBar()
         )
 
-    if args.profiling:
-        model.callbacks.append(lbann.CallbackProfiler())
+    profiler = create_profile_callback(args)
```
Comment (tbennun) [inline applications/nlp/transformer/train.py:229]: where is this imported from?

**Review by tbennun — CHANGES_REQUESTED** [review]
One comment on the Python side (please use some module path for that function)

**Review by fiedorowicz1 — APPROVED** [review]
Looks good other than @tbennun's comment

File: applications/nlp/transformer/train.py (hunk 1 above)
Comment (ndryden) [reply applications/nlp/transformer/train.py]: Fixed. Thanks!

**Review by tbennun — APPROVED** [review]

**Pull request merged**